crates/fs/private/namefilter/namefilter.rs

```diff
@@ -0,0 +1,75 @@
+use sha3::{
+    digest::{ExtendableOutput, Update, XofReader},
+    Shake256,
+};
+
+use super::BloomFilter;
+
+//--------------------------------------------------------------------------------------------------
+// Constants
+//--------------------------------------------------------------------------------------------------
+
+pub const SATURATION_THRESHOLD: usize = 1019;
+
+//--------------------------------------------------------------------------------------------------
+// Type Definitions
+//--------------------------------------------------------------------------------------------------
+
+/// Namefilters are 2048-bit bloom filters.
+///
+/// In WNFS they represent the identity of a file or directory, doubling as a store for checking the ancestor of the file or directory.
+pub type Namefilter = BloomFilter<256, 30>;
+
+//--------------------------------------------------------------------------------------------------
+// Implementations
+//--------------------------------------------------------------------------------------------------
+
+impl Namefilter {
+    /// Creates a saturated namefilter.
+    pub fn new_saturated() -> Self {
+        let mut filter = Self::new();
+        filter.saturate();
+        filter
+    }
+
+    /// Adds hashes to filter until is is over the saturation threshold.
+    fn saturate(&mut self) {
+        let mut xof = {
+            let mut h = Shake256::default();
+            h.update(self.as_bytes());
+            h.finalize_xof()
+        };
+
+        while self.count_ones() <= SATURATION_THRESHOLD {
```

Comment:
Yeah, I know this is confusing, but we actually want to stay *slightly below* the saturation threshold. So once we realize that we're above the threshold, we need to "undo" the last step.